File: Proyecto Final Progra/java-benchmark/Benchmark.java
import java.io.FileWriter;
import java.util.HashMap;
import java.util.TreeMap;

public class Benchmark {

    public static void run() {

        try {

            FileWriter out =
                new FileWriter(
                    "java_results.csv"
                );

            out.write(
                "language,operation,structure,records,time_ms\n"
            );

            int RECORDS = 10000;

            // =========================
            // HASHMAP INSERT
            // =========================

            HashMap<Integer, Student>
                hashMap =
                    new HashMap<>();

            long start1 =
                System.nanoTime();

            for(int i = 0; i < RECORDS; i++) {

                hashMap.put(
                    i,
                    new Student(
                        i,
                        "Student"
                    )
                );
            }

            long end1 =
                System.nanoTime();

            double insertHash =
                (end1 - start1)
                / 1000000.0;

            out.write(
                "Java,insert,HashMap,"
                + RECORDS
                + ","
                + insertHash
                + "\n"
            );

            // =========================
            // HASHMAP SEARCH
            // =========================

            long start2 =
                System.nanoTime();

            for(int i = 0; i < RECORDS; i++) {

                hashMap.get(i);
            }

            long end2 =
                System.nanoTime();

            double searchHash =
                (end2 - start2)
                / 1000000.0;

            out.write(
                "Java,search,HashMap,"
                + RECORDS
                + ","
                + searchHash
                + "\n"
            );

            // =========================
            // TREEMAP INSERT
            // =========================

            TreeMap<Integer, Student>
                treeMap =
                    new TreeMap<>();

            long start3 =
                System.nanoTime();

            for(int i = 0; i < RECORDS; i++) {

                treeMap.put(
                    i,
                    new Student(
                        i,
                        "Student"
                    )
                );
            }

            long end3 =
                System.nanoTime();

            double insertTree =
                (end3 - start3)
                / 1000000.0;

            out.write(
                "Java,insert,TreeMap,"
                + RECORDS
                + ","
                + insertTree
                + "\n"
            );

            // =========================
            // TREEMAP TRAVERSAL
            // =========================

            long start4 =
                System.nanoTime();

            treeMap.values();

            long end4 =
                System.nanoTime();

            double traversal =
                (end4 - start4)
                / 1000000.0;

            out.write(
                "Java,traversal,TreeMap,"
                + RECORDS
                + ","
                + traversal
                + "\n"
            );

            out.close();

            System.out.println(
                "java_results.csv generado"
            );

        } catch(Exception e) {

            System.out.println(e);
        }
    }
}

File: Proyecto Final Progra/cpp-system/Benchmark.cpp
// ============================
// Benchmark.cpp
// ============================

#include "Benchmark.h"

#include <fstream>
#include <chrono>

using namespace chrono;

void Benchmark::run(
    HashTable& hash,
    AVLTree& avl,
    Graph& graph
) {

    ofstream out(
        "cpp_results.csv"
    );

    out
    << "language,operation,structure,records,time_ms\n";

    const int RECORDS = 10000;

    Student* students[RECORDS];

    for(int i = 0; i < RECORDS; i++) {

        students[i] =
            new Student(
                50000 + i,
                "Benchmark",
                "Systems",
                5,
                80,
                i
            );
    }

    auto start1 =
        high_resolution_clock::now();

    for(int i = 0; i < RECORDS; i++) {

        hash.insert(students[i]);
    }

    auto end1 =
        high_resolution_clock::now();

    double hashInsert =
        duration<double, milli>(
            end1 - start1
        ).count();

    out
    << "C++,insert,HashTable,"
    << RECORDS
    << ","
    << hashInsert
    << "\n";

    auto start2 =
        high_resolution_clock::now();

    for(int i = 0; i < RECORDS; i++) {

        hash.search(50000 + i);
    }

    auto end2 =
        high_resolution_clock::now();

    double hashSearch =
        duration<double, milli>(
            end2 - start2
        ).count();

    out
    << "C++,search,HashTable,"
    << RECORDS
    << ","
    << hashSearch
    << "\n";

    auto start3 =
        high_resolution_clock::now();

    for(int i = 0; i < RECORDS; i++) {

        avl.insert(students[i]);
    }

    auto end3 =
        high_resolution_clock::now();

    double avlInsert =
        duration<double, milli>(
            end3 - start3
        ).count();

    out
    << "C++,insert,AVL,"
    << RECORDS
    << ","
    << avlInsert
    << "\n";

    auto start4 =
        high_resolution_clock::now();

    avl.inorder();

    auto end4 =
        high_resolution_clock::now();

    double avlTraversal =
        duration<double, milli>(
            end4 - start4
        ).count();

    out
    << "C++,traversal,AVL,"
    << RECORDS
    << ","
    << avlTraversal
    << "\n";

    for(int i = 0; i < RECORDS - 1; i++) {

        graph.addConnection(
            i,
            i + 1
        );
    }

    auto start5 =
        high_resolution_clock::now();

    graph.BFS(0);

    auto end5 =
        high_resolution_clock::now();

    double bfsTime =
        duration<double, milli>(
            end5 - start5
        ).count();

    out
    << "C++,bfs,Graph,"
    << RECORDS
    << ","
    << bfsTime
    << "\n";

    auto start6 =
        high_resolution_clock::now();

    graph.DFS(0);

    auto end6 =
        high_resolution_clock::now();

    double dfsTime =
        duration<double, milli>(
            end6 - start6
        ).count();

    out
    << "C++,dfs,Graph,"
    << RECORDS
    << ","
    << dfsTime
    << "\n";

    out.close();

    cout
        << "cpp_results.csv generado\n";
}